### ISSUE TITLE
Truly alternative backend

### ISSUE THREAD

**Issue by kvark** [szeged/webrender#145]
Thinking about the future upstreaming of that branch, it would be much safer to transition if we could easily flip a (run-time?) switch to go back and forth between the old GL implementation and the new HAL. This would require maintaining (for a limited time frame) both the old and new device code, having the common (device API) parts moved outside for connect with renderer and such.

Filing this issue to raise that question. Please let me know what you think about the need for this hybrid solution as well as the feasibility of it.

In relation to the proposal, it seems like we'd benefit from more space inside the device module if we move it into a folder, thus potentially ending up with:
-  `device/gl.rs`
-  `device/hal.rs`
-  `device/api.rs`

If that seems feasible, I'll prioritize starting this move upstream, and then we'll update this branch as a follow-up.

**Comment by dati91** [comment]
Hmm, an interesting proposal.
I totally agree on the
> it would be much safer to transition

part, that is a great point.

When we started working on this, our goal was to try to change as minimal as we can on the (device) api, so the upstream rebases could go as smooth as possible. (next one is coming up soon so 🤞)

The two main things that come to mind first is that
- the renderer "knows" that it uses a gl backend, and uses things like "vao"s
- and also it calls actual gl calls

After removing those, we just call the (mostly) same device.rs calls, with a different implementation under the hood of course.

So if we want to use a "common api" those are the things that should be abstracted away first.

Note: I didn't checked/saw the latest WR's code in a while, so there is a chance that some of these issues are already solved, or even more came in :)

**Comment by kvark** [comment]
> the renderer "knows" that it uses a gl backend, and uses things like "vao"s

This is certainly unfortunate. Do you think we could upstream some sort of PSO notion, so that the device internally treats PSO = VAO + shader for the old GL backend?

> and also it calls actual gl calls

I've been addressing those, and will continue doing that. Any GL access outside of `device` module in WR can be considered an upstream bug already.

**Comment by dati91** [comment]
So I think it is time to revive this discussion.

We already started to make our life easier with #150 

The main topics:
- matching the device.rs calls, so we can use the `same` api
- renderer initialization
- shader/program creation
- wrench

#
> Do you think we could upstream some sort of PSO notion, so that the device internally treats PSO = VAO + shader for the old GL backend?

So the easier "way out" for now would be that we create dummy structures for those (`PBO`, `VAO`, `VBO`, `VertexUsageHint`, `CustomVAO`).
If we do this, we match the device api without changing the signature of the functions.
I know that this is not the greatest solution, but we can always come back later to add a structure like `PSO = VAO + shader`. And when we do, we can change both at the same time, and more likely we will handle any unseen problems.

The same goes for the `GpuProfiler`. We can make an empty stub for that and later come back and fill in the actual implementation.

#
There are differences in the renderer::new call.
On OpenGL, we need the gl context, On gfx, we need an adapter and a surface.
We can create an enum to store these and to have the same new call signature. Probably.
Both of these inputs go to the device struct.

Also we use `Device<B>` and the original is `Device`. I hope the `PhantomData` magic can help with that, but I didn't look into that just yet.

#
We create our shaders in build time and load them in run time. And there are gl specific things in `shade.rs` file as well.
The good news is that the public part of `shade.rs` is almost the same (there is a gl_type param when calling new).
The simplest way to handle this to make an api like the device does(/will do).

Also we can do the build time shader concatenation for the gl as well, because we use (mostly) the same code what the original concatenation does but in the build.rs.

#
Wrench uses glutin, we need winit. I guess this will change in upstream in the (near?) future.
It supports headless context as well. We are not there yet. But we will need that for Servo's wpt (iirc). Vulkan can do that, but I'm not sure about metal and dx12, will have to look into that.

**Comment by kvark** [comment]
> So the easier "way out" for now would be that we create dummy structures for those (PBO, VAO, VBO, VertexUsageHint, CustomVAO).

If that's easy for you then fine. I feel that upstreaming the notion of PSO would be more beneficial though, since it would cater to the new backend better than the old one.

> Wrench uses glutin, we need winit. I guess this will change in upstream in the (near?) future.

We do need to switch as much of `wrench` as possible to use `winit` (e.g. for events) instead of `glutin` (similarly to what we did for Servo) - upstream. Then we'd leave GL specific parts behind a feature.

> It supports headless context as well. We are not there yet. But we will need that for Servo's wpt (iirc). Vulkan can do that, but I'm not sure about metal and dx12, will have to look into that.

Beware of the headless contexts - the term is ambiguous. In Glutin, a headless context is backed by OSMesa, so it's actually a software context. If we are simply talking about contexts without display surfaces associated, then all next-gen backends natively support that. However, we'll not be able to run those on CI (for Servo WPT or WR reftests).

**Comment by dati91** [comment]
We managed to [achieve a state](https://github.com/dati91/webrender/tree/alternative_backend) where our code and the original can live together, in a bit of ugly way, but i think it is a good starting point.

Our approach is the similar to proposed one, but in a very simplified way.
We create a folder for all the files which requires two implementation.
For now, we have the implementations in separate files: `gl.rs` and `gfx.rs`.
And we choose between them in a [`mod.rs`](https://github.com/dati91/webrender/blob/alternative_backend/webrender/src/device/mod.rs) file, depending on the features.
This can only work if we have the same API (or a very similar one with a [minor difference](https://github.com/dati91/webrender/blob/alternative_backend/webrender/src/renderer.rs#L1516)).

This has the advantage that we don't have to modify the original files (only move them) to have a working prototype. But in reality we actually do have to modify them e.g. add a `PhantomData` to have the same structure signature.

But a better way to do this would be using an API trait, as proposed before, where we extract the publicly available functions from the original implementation, and implement them in both version (gl and gfx).
This could be our next iteration in this topic.

The main advantage of this prototype that it showed us where the code have to change:

Renderer, Device, Shaders, etc. structures that in touch with the device have to be generic (because of the hal::Backend trait). I really tried, but i just  can't seem to find a way get around it.
The gfx part can use the hal::Backend Trait, and the OpenGL part can use a PhantomData. This also means that the implementer app (Servo, Wrench, ... ) has to [know about gfx](https://github.com/dati91/webrender/blob/alternative_backend/wrench/src/wrench.rs#L151), even if it not used (empty crate).

The `device`, `shade`, and `query` folders were created, because these differ a lot. (And the `debug_render` at the moment because our program creation differs from the original.) These files should have a common API.

The shader compiling process is currently separated. The original implementation compiles it in run time, and we do it in build time. Our current approach is to [call](https://github.com/dati91/webrender/blob/alternative_backend/webrender/build.rs#L87) into [our compilation](https://github.com/dati91/webrender/blob/alternative_backend/webrender/build_gfx.rs), guarded with a feature cfg. This also means that we have a shaders.ron file even if the gl backend is used.
The main difference is that we need to concatenate the shaders and compile it in build time. The upstream version (easily?) could use the build time concatenation approach, and compile it in runtime.

Regarding the previous comment:

> all next-gen backends natively support that

Great!

>  However, we'll not be able to run those on CI (for Servo WPT or WR reftests).

This is not that great :/

**Comment by zakorgy** [comment]
I picked up the work @dati91 started [here](https://github.com/dati91/webrender/tree/alternative_backend), and continue on this [branch](https://github.com/zakorgy/webrender/commits/tab).

What is done:

 - We have a `gfx` feature in WR, which can enable our backend. Without that feature the default implementation is used. We currently use `cfg`-s to achieve this, but in the future we plan to add a common trait which the different backends implement (as it was mentioned above).

- Wrench has fetaures to run with vulkan and dx12 (`vulkand` and `dx12` features). Without any of those features it runs with the original implementation using gleam.

 WIP:

- [ ] I have moved the examples from the `webrender/examples` folder one level up into `webrender_examples` folder where we have features for the different backends. The features will work the same way as described in Wrench. The `yuv` and `frame_output` examples are heavily `gleam` dependent. Until we figure out how to replace those `gl` parts, we can config these to run with `gleam` only.

What needs to be done:

 - [ ] Create a common trait which the different `Device`s implement, and replace the usage of `Device` in `Renderer` with that.

- [ ] Make the `DebugRenderer` work with both device implementations.

- [ ] Enable the angle shader validation. This was previously mentioned in https://github.com/szeged/webrender/issues/150#issuecomment-388843262
